guardian/cdk
# feat: add first attempt at experimental error budget alarm

## What does this change?

This PR adds a new experimental construct which aims to create alarms that will notify teams when they are at risk of missing their SLO targets.

The alarm configuration that we create follows the recommendations in the [SRE Workbook](https://sre.google/workbook/alerting-on-slos/) to implement `Multiwindow, Multi-Burn-Rate Alerts`. There is also another nice explanation of this alerting strategy [here](https://developers.soundcloud.com/blog/alerting-on-slos).
 
The CloudWatch implementation of this strategy creates nine separate CloudWatch alarms. The main alarms that users need to care about are the three [composite alarms](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Create_Composite_Alarm.html). These track different burn rates of error budget consumption: 

* fast (e.g. complete outages)
* medium (e.g. moderate error percentage over hours, rather than minutes) 
* slow (e.g. a small percentage of errors sustained over a much longer period)

Multiple burn rates are used to ensure that we are able to pick up all scenarios which could threaten the SLO, without interrupting developers with an unnecessary level of urgency:

![image](https://user-images.githubusercontent.com/19384074/198985926-e6f38b03-aa4b-48a5-83c1-f2d338c4ceaf.png)
_The image above is taken from Google's `Alerting on SLOs` chapter._

Each composite alarm creates two child alarms; these are used to support multiple windows (or periods in CloudWatch terminology). A composite alarm is configured to send notifications if both of its child alarms are in an `ALARM` state. The child alarms themselves have no actions/notifications configured and users should not need to interact with them directly.  Here's an example of how that looks in the AWS console:

![image](https://user-images.githubusercontent.com/19384074/199013792-e7c3adbd-63cc-4b15-bb2a-8eaf6452b8aa.png)

Multiple windows are helpful for accuracy and reset time. The long window helps us to avoid sending an alert in scenarios where no intervention is required, for example a very brief spike in errors that is resolved automatically (e.g. when AWS replaces a 'bad' EC2 instance). The short window helps to ensure that the alert is reset (or moved back into `OK` state, in CloudWatch terminology) promptly once a problem has been resolved:

![image](https://user-images.githubusercontent.com/19384074/198985995-bd54e737-082d-43d0-9e4c-bc32c60be5ed.png)
_The image above is taken from Google's `Alerting on SLOs` chapter._

The composite alarms have an awareness of priority, meaning that low priority alarm notifications are [suppressed](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Create_Composite_Alarm.html#Create_Composite_Alarm_Suppression) if a higher priority alarm is already firing. That is, if you receive an alert about fast error budget consumption, you should not receive an alert from the medium or slow versions of the alarm. Here's an example of how that looks in the AWS console:

![image](https://user-images.githubusercontent.com/19384074/199013923-63d418a0-4706-4095-8b75-76a3e6584a4c.png)

Although the generated alarm configuration is quite complex most of this is abstracted away. Developers working on `cdk` code need only supply high level configuration like this:

```typescript
new GuErrorBudgetAlarmExperimental(stack, {
    sloName: "MapiFrontsAvailability",
    sloTarget: 0.999,
    badEvents: new Metric({ metricName: "HttpErrors", namespace: "TestLoadBalancerMetrics" }),
    validEvents: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
    snsTopicNameForAlerts: "test-sns-topic",
});
```

## How to test

I have (manually) tested this alert configuration using the `cdk-playground` project to confirm that it works as expected. I've added a couple of snapshot tests to document how this works and keep a record of a working version of the configuration. We'll need to continue to test this with real services so that we can see how it behaves with real traffic patterns and real SLOs etc.

## How can we measure success?

We will attempt to use this construct with the MSS team over the next couple of months. We hope that this approach to alerting will prove to be useful. If we're correct about that (and we decide that CloudWatch is an appropriate tool for this task) then we will look to 'productionise' this construct so that it can be used more widely. If it is not helpful (or if CloudWatch doesn't seem like the best tool for the job), we'll just delete this construct and help MSS to pick up the preferred solution instead.

## Have we considered potential risks?

I think this is pretty low risk. We are experimenting with a new technique and are therefore using an experimental construct, as per [this ADR](https://github.com/guardian/cdk/blob/main/docs/architecture-decision-records/005-x01-package-structure.md).

## Checklist


[^1]: Consider whether this is something that will mean changes to projects that have already been migrated, or to the CDK CLI tool. If changes are required, consider adding a checklist here and/or linking to related PRs.
[^2]: If you are adding a new construct or pattern, has new documentation been added? If you are amending defaults or changing behaviour, are the existing docs still valid?
